### ISSUE TITLE
Breaking change in filter functionality between framework 2.7 and 3.11

### ISSUE THREAD

**Issue by jbengaard** [nunit/nunit-console#606]
We have a filter, --where="!(test != <something>)" which returned the appropriate tests on nunit framework 2,7.0.0. On 3.11.0.0, it always returns an empty set of tests.
This behavior was observed using NUnit Console Runner 3.10.0.

I attached a small example solution containing a few tests using 2.7,0.0 and a copy of the tests using 3.11.0.0. I also attached the output from running the tests with the same filter.

[nunit2example_1.zip](https://github.com/nunit/nunit-console/files/3080284/nunit2example_1.zip)
![Console output](https://user-images.githubusercontent.com/46798085/56135558-4e4bd500-5f91-11e9-8632-acbf752a52bb.PNG)

**Comment by jbengaard** [comment]
I tried to use nunit framework  v3.0.5797 and console runner v3.0.5797,
with the same result.

Den man. 15. apr. 2019 kl. 16.57 skrev CharliePoole <
notifications@github.com>:

**Comment by CharliePoole** [comment]
Sounds like the problem has been there for a while then. Can you avoid that specific syntax as a workaround?

**Comment by CharliePoole** [comment]
This is probably a framework error, but someone should confirm before transferring it I think.

**Comment by jbengaard** [comment]
Unfortunately, we have some tests running on framework 2.7 that need that specific type of filter. We have not found a workaround yet.

**Comment by CharliePoole** [comment]
You should be aware that `!(test != X)` is not the same as `test == X` if X happens to be marked `Explicit`.

That is, a top level not-filter isn't allowed to select an Explicit test. For non-explicit tests, the normal equivalence is intended to apply.

**Comment by cyanite** [comment]
> You should be aware that `!(test != X)` is not the same as `test == X` if X happens to be marked `Explicit`.

Well, `!(test != X)` is not the same as `test == X` in general, since in our case (I am a collegue of @jbengaard), I get 0 tests run with `--where="!(test!='Some.TopLevel.Namespace')"` and all tests run with the De Morgan equivalent `--where="test='Some.TopLevel.Namespace'"`.

**Comment by CharliePoole** [comment]
Yes, I think we understand the problem you are seeing. I am only pointing out one small exception where Explicit is involved.

Someone will need to verify that the generated filter is correct before possibly moving the issue to the framework repo.

**Comment by CharliePoole** [comment]
I can confirm this bug. See build at https://ci.appveyor.com/project/CharliePoole/nunit-console/builds/24263259

In this branch I added tests to

1. Verifiy that the filter is generated correctly, for example, `!(test != X)` gives `<not><not><test>X</test></not></not>`. These tests pass.

2. Verify that a filter beginning with `<not><not>` performs as expected. Those tests fail.

I'm going to go ahead and prepare a PR to add the first tests, but not the second ones. We can't have those tests since they will fail unless we first update the framework. We can either leave them out entirely or comment them until the framework version used is updated.

**Comment by CharliePoole** [comment]
@nunit/engine-team I'm waiting for someone to confirm my view that this issue should be moved to the framework repo before doing it. See the above build.

**Comment by jnm2** [comment]
@CharliePoole I agree.

**Comment by ChrisMaddock** [comment]
Me to!
